pyamg/strength.py

```diff
@@ -175,37 +182,81 @@ def classical_strength_of_connection(A, theta=0.0, norm='abs'):
     else:
         blocksize = 1
 
-    if not sparse.isspmatrix_csr(A):
-        warn('Implicit conversion of A to csr', sparse.SparseEfficiencyWarning)
-        A = sparse.csr_matrix(A)
-
     if (theta < 0 or theta > 1):
         raise ValueError('expected theta in [0,1]')
 
-    Sp = np.empty_like(A.indptr)
-    Sj = np.empty_like(A.indices)
-    Sx = np.empty_like(A.data)
+    # Block structure considered before computing SOC
+    if (block == 'block') or sparse.isspmatrix_bsr(A):
+        R, C = A.blocksize
+        if (R != C) or (R < 1):
+            raise ValueError('Matrix must have square blocks')
+
+        N = int(A.shape[0] / R)
+
+        # SOC based on maximum absolute value element in each block
+        if norm == 'abs':
+            data = np.max(np.max(np.abs(A.data),axis=1),axis=1)
+        # SOC based on hard minimum of entry in each off-diagonal block
+        elif norm == 'min':
+            data = np.min(np.min(A.data,axis=1),axis=1)
+        # SOC based on Frobenius norms of blocks
+        elif norm == 'fro':
+            data = (np.conjugate(A.data) * A.data).reshape(-1, R*C).sum(axis=1)
+        else:
+            raise ValueError("Invalid choice of norm.")
+
+        data[np.abs(data)<1e-16] = 0.0
+        S_indptr = np.empty_like(A.indptr)
+        S_indices = np.empty_like(A.indices)
+        S_data = np.empty_like(data)
+
```

Comment:
Just to make sure I understand things, 
data will be positive for 'fro' and 'abs' case, but that's OK, because the classical_soc_abs() routine only looks at magnitude
but then, data will be negative for 'min, but that's OK, that's the normal case.

Reply:
yep exactly

Reply:
Cool